File: Semester-3/Computer-Networks/Templates/UDP-server.c
#include <netinet/in.h>
#include <stdio.h>
#include <strings.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char* argv[]) {
    int server_socket;
    struct sockaddr_in server, from;
    int cod;
    char buffer[1024];

    server_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (server_socket < 0) {
        printf("Error on creating socket!\n");
        return 1;
    }

    bzero(&server, sizeof(server));
    server.sin_family = AF_INET;
    server.sin_port = htons(atoi(argv[1]));
    server.sin_addr.s_addr = INADDR_ANY;

    cod = bind(server_socket, (struct sockaddr*)&server, sizeof(server));
    if (cod < 0) {
        printf("Bind error!\n");
        return 1;
    }

    // receive/send a string

    while (1) {
        cod = recvfrom(server_socket, buffer, 1024, 0, (struct sockaddr*)&from, sizeof(from));
        if (cod < 0) {
            printf("Receive error!\n");
            continue;
        }

        cod = sendto(server_socket, buffer, sizeof(buffer), 0, (struct sockaddr*)&from, sizeof(from));
        if (cod < 0) {
            printf("Send error!\n");
            continue;
        }
    }

    close(server_socket);



    return 0;
}

File: Semester-3/Computer-Networks/Templates/UDP-server.py
import socket
import struct
import threading

def worker(s):
    # receive number
    n, addr = s.recvfrom(2)
    n = struct.unpack("!H", n1)[0]

    # send number
    res = s.sendto(struct.pack("!H", n), addr) # !!!! SEND TO ADDR, PORT IS DIFFERENT

    # close socket
    s.close()


if __name__ == "__main__":
    try:
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        server_socket.bind(("ip addr", port))
        print("Waiting for client!")
    except socket.error as err:
        print("Error: " + err.strerror)
        exit(-1)

    t = threading.Thread(target=worker, args=(server_socket,))
    t.start()
    t.join()